hazelcast/include/hazelcast/client/connection/Connection.h

```diff
@@ -57,74 +63,107 @@ namespace hazelcast {
 
                 void setRemoteEndpoint(const Address& remoteEndpoint);
 
                 Socket& getSocket();
 
-                std::auto_ptr<protocol::ClientMessage> sendAndReceive(protocol::ClientMessage &clientMessage);
-
                 ReadHandler& getReadHandler();
 
                 WriteHandler& getWriteHandler();
 
-                void setAsOwnerConnection(bool isOwnerConnection);
+                bool isHeartBeating();
 
-                void writeBlocking(protocol::ClientMessage &packet);
+                void onHeartbeatFailed();
 
-                std::auto_ptr<protocol::ClientMessage> readBlocking();
+                void onHeartbeatResumed();
 
-                bool isHeartBeating();
+                void onHeartbeatReceived();
 
-                void heartBeatingFailed();
+                void onHeartbeatRequested();
 
-                void heartBeatingSucceed();
+                bool isAuthenticatedAsOwner() const;
 
-                bool isOwnerConnection() const;
+                void setIsAuthenticatedAsOwner();
 
-                virtual void handleMessage(connection::Connection &connection, std::auto_ptr<protocol::ClientMessage> message);
+                virtual void handleClientMessage(const boost::shared_ptr<Connection> &connection,
+                                                 std::auto_ptr<protocol::ClientMessage> &message);
 
                 int getConnectionId() const;
 
-                void setConnectionId(int connectionId);
+                bool isAlive();
+
+                int64_t lastReadTimeMillis();
+
+                const std::string &getCloseReason() const;
+
+                void incrementPendingPacketCount();
+
+                void decrementPendingPacketCount();
+
+                int32_t getPendingPacketCount();
+
+                bool operator==(const Connection &rhs) const;
+
+                bool operator!=(const Connection &rhs) const;
+
+                const std::string &getConnectedServerVersionString() const;
+
+                void setConnectedServerVersion(const std::string &connectedServerVersionString);
+
+                std::auto_ptr<Address> getLocalSocketAddress() const;
+
+                int getConnectedServerVersion() const;
+
+                friend std::ostream &operator<<(std::ostream &os, const Connection &connection);
 
-                util::Atomic<time_t> lastRead;
-                util::AtomicBoolean live;
             private:
+                util::Atomic<int64_t> closedTimeMillis;
+                util::Atomic<int64_t> lastHeartbeatRequestedMillis;
+                // the time in millis the last heartbeat was received. 0 indicates that no heartbeat has ever been received.
+                util::Atomic<int64_t> lastHeartbeatReceivedMillis;
                 spi::ClientContext& clientContext;
-                spi::InvocationService& invocationService;
+                spi::ClientInvocationService &invocationService;
```

Comment:
Instead of keeping this invocation service, we could keep this as IMessageHandler. It is the only method we use. 
Also ListenerService can be made IMessageHandler, and kept here similarly.